grpc-sys/src/lib.rs

```diff
@@ -348,20 +351,65 @@ pub struct GrpcMetadataArray {
     pub metadata: *mut GrpcMetadata,
 }
 
-#[cfg_attr(feature = "cargo-clippy", allow(unreadable_literal))]
-pub const GRPC_INITIAL_METADATA_IDEMPOTENT_REQUEST: uint32_t = 0x00000010;
-#[cfg_attr(feature = "cargo-clippy", allow(unreadable_literal))]
-pub const GRPC_INITIAL_METADATA_WAIT_FOR_READY: uint32_t = 0x00000020;
-#[cfg_attr(feature = "cargo-clippy", allow(unreadable_literal))]
-pub const GRPC_INITIAL_METADATA_CACHEABLE_REQUEST: uint32_t = 0x00000040;
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct GrpcSliceRefCounted {
+    bytes: *mut uint8_t,
+    length: size_t,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct GrpcSliceInlined {
+    length: uint8_t,
+    // TODO: use size_of when it becomes a const function.
+    #[cfg(target_pointer_width = "64")]
+    bytes: [uint8_t; 23],
+    #[cfg(target_pointer_width = "32")]
+    bytes: [uint8_t; 11],
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub union GrpcSliceData {
+    ref_counted: GrpcSliceRefCounted,
+    inlined: GrpcSliceInlined,
+}
+
+pub enum GrpcSliceRefCount {}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
```

Comment:
Seems that some codes are depending on the deriving of `Copy` and `Clone`.

Reply:
For example, what does it mean when calling `Clone` on `GrpcSlice`? Does the refcount need to be changed?

Reply:
No, this is just a reference to a C object. Cloning a reference is quite common.

This struct is just for forwarding some C stuffs, AFAIK (if I understand your code correctly).

Reply:
> This struct is just for forwarding some C stuffs

This means you never create `GrpcSlice` yourself, it's created on C-side. You never destruct it yourself, it's destructed on C-side. Your snippet has initialized a `GrpcSlice` on Rust-side, it's absolutely and obviously unsafe. But we've never did that, do we?

This is why `grpc_slice_unref` is [not used](https://github.com/pingcap/grpc-rs/search?q=grpc_slice_unref&unscoped_q=grpc_slice_unref) on Rust-side.

Reply:
Sorry I've just read the pr again and I found that it's used [here](https://github.com/pingcap/grpc-rs/pull/222/files#diff-8c8308f2c854f39735f331e70f94d8d2R184). But that looks safe to me anyway

Reply:
Please don't just focus on how grpcio uses the API. grpcio-sys is a public crate, people is free to depend on the ABI and implements what they want. Hence `Clone` and `Copy` need to be considered carefully here.

Reply:
@BusyJay This is a separated task, I'm not doing this in this PR.

Reply:
Why is it a separated task? Is `Clone` and `Copy` necessary? What would happen if just removing them?

Reply:
You'll get compilation error due o implicit copy is no longer possible

Reply:
I think the conversation should be not resolved. I think we need to remove `Clone` and `Copy` derive at least. Any updates?

Reply:
@BusyJay @hicqu Currently the impl of `MessageReader` is C-style, its maintainability is bad but its logic is correct. I'll refactor this part in a separate PR to avoid taking too much time by further code review&discussion.

BTW PTAL